### PR TITLE
[keras] Fix AUC.result returning 0 for single-class ROC inputs

### LINKED ISSUE #41081: Zero AUROC when validation set is only comprised of positive labels
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Tested on Ubuntu 18.04 and MacOS Catalina 10.15.5
- TensorFlow installed from (source or binary): installed from Conda (tensorflow-gpu) on Ubuntu and from pip on MacOS
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.7.6 (on both machines)
- CUDA/cuDNN version: 11.0 (on Ubuntu machine)
- GPU model and memory: 4 NVIDIA V100 GPUs

**Describe the current behavior**
The AUROC is zero when there are no negatives in the validation set. This may lead users to believe that the model is not performing correctly, while it is just an error related to a misuse of the AUROC metric.

**Describe the expected behavior**
The AUROC should be NaN when no negatives are present in the validation set. This way the user is aware that there is something wrong with the metric that is being used, with the possibility to add a warning since there is a 0/0 somewhere in the computation of the AUROC.

**Standalone code to reproduce the issue**
```python
from tensorflow.keras.metrics import AUC
metric = AUC()
metric.update_state([1, 1, 1], [1, 0.5, 0.3])
metric.result()
```

I believe that similar issues also happen for other metrics, as uncanny values were displayed also for Recall and Precision, but I have not identified a simple reproducible example.

### PR DESCRIPTION
**Pull request by Ijtihed**
When y_true contains only one class (all-positive or all-negative) ROC-AUC is undefined. AUC.result() returned 0.0 because div_no_nan silently maps 0/0 to 0 making all fp-rate differences zero in the Riemann sum.

Checks true_positives[0] and false_positives[0] at threshold -epsilon (where all examples are predicted positive) and returns NaN if either class is absent. 

Fixes #41081

### CODE REVIEW

**Comment by Ijtihed** [comment]
cc @ratgr @keerthanakadiri gh-121465

**Review by ratgr — CHANGES_REQUESTED** [review]
Thank you for updating the PR and adding comprehensive unit tests! The mathematical logic for returning `NaN` on single-class ROC inputs and excluding them from multi-label averages looks solid.

The GitHub Actions **PyLint** check failed due to a minor line-length violation in `tensorflow/python/keras/metrics_test.py`:
- `Line too long (81/80)` on lines with `y_pred = [...]` and comment `# Weighted mean over defined labels...`.

Could you please format those lines to stay within 80 characters? Once that's fixed, this PR will be ready to merge.

**Comment by Ijtihed** [comment]
> Thank you for updating the PR and adding comprehensive unit tests! The mathematical logic for returning `NaN` on single-class ROC inputs and excluding them from multi-label averages looks solid.
> 
> The GitHub Actions **PyLint** check failed due to a minor line-length violation in `tensorflow/python/keras/metrics_test.py`:
> 
> * `Line too long (81/80)` on lines with `y_pred = [...]` and comment `# Weighted mean over defined labels...`.
> 
> Could you please format those lines to stay within 80 characters? Once that's fixed, this PR will be ready to merge.

thanks for the review, done!

**Review by ratgr — APPROVED** [review]
Thank you @Ijtihed for addressing the review comments and updating the formatting!

Everything looks great:
- All GitHub CI checks (including PyLint) are now passing cleanly.
- The mathematical logic for returning `NaN` on single-class ROC inputs and excluding them from multi-label macro/weighted averages is sound and well-tested.
- Code style and line lengths comply with repository guidelines.

PR is approved!

**Comment by Ijtihed** [comment]
Pinging on this

**Comment by ratgr** [comment]
Thank you for your contribution! This PR is currently undergoing some additional internal validation checks. We will let you know if any further action is needed from your side.

**Review by ratgr — CHANGES_REQUESTED** [review]
Hi @Ijtihed,
Thank you very much for submitting this pull request and for your effort in improving the codebase!
We apologize for the inconvenience, but we are unable to merge this pull request directly in the `tensorflow/tensorflow` repository. The file(s) modified in this PR:
* `tensorflow/python/keras/BUILD` → `keras/BUILD` in [keras-team/keras](https://github.com/keras-team/keras) (or [keras-team/tf-keras](https://github.com/keras-team/tf-keras))
* `tensorflow/python/keras/metrics.py` → `keras/metrics.py` in [keras-team/keras](https://github.com/keras-team/keras) (or [keras-team/tf-keras](https://github.com/keras-team/tf-keras))
* `tensorflow/python/keras/metrics_test.py` → `keras/metrics_test.py` in [keras-team/keras](https://github.com/keras-team/keras) (or [keras-team/tf-keras](https://github.com/keras-team/tf-keras))
are maintained and developed in the upstream [Keras](https://github.com/keras-team/keras) repository and mirrored into TensorFlow. Because these components belong to a separate upstream repository, changes made directly here cannot be imported into the internal codebase (which is why the `import/copybara — Import didn't affect any internal file` check was reported).
Could you please re-submit this pull request against the [keras-team/keras](https://github.com/keras-team/keras) repository instead?
### Suggested Next Steps:
1. Fork and clone the upstream repository: [keras-team/keras](https://github.com/keras-team/keras).
2. Create a new branch and apply your changes to the corresponding file(s).
3. Open a new pull request targeting the main branch of [keras-team/keras](https://github.com/keras-team/keras).
We sincerely apologize for the redirection and the extra step, and we truly appreciate your contribution and understanding!

**Comment by ratgr** [comment]
Hi @Ijtihed, thank you so much for taking the time to investigate this issue and put together this fix!

The `tensorflow/python/keras` directory in `tensorflow/tensorflow` is a frozen legacy snapshot preserved only for internal backwards compatibility, and it is no longer accepting general bug fixes or metric updates (only critical security or build-breakage fixes). All active Keras development and maintenance now takes place in the standalone Keras repositories.

Because of this policy, we are going to close this PR here, but we'd love for you to contribute this fix upstream to **[`keras-team/keras`](https://github.com/keras-team/keras)** (or **[`keras-team/tf-keras`](https://github.com/keras-team/tf-keras)** for the TensorFlow 2 legacy Keras package) if the issue is still present there.

Thanks again for your contribution and understanding!
